include/alp/reference/matrix.hpp

```diff
@@ -432,6 +439,23 @@ namespace alp {
 				std::pair< size_t, size_t > dims() const noexcept {
 					return static_cast< const DerivedMatrix * >( this )->_dims();
 				}
+
+				template< typename MatrixType >
+				friend const typename MatrixType::value_type &getReference( const MatrixType &A, const size_t i, const size_t j );
+
+				template< typename MatrixType >
+				friend typename MatrixType::value_type &getReference( MatrixType &A, const size_t i, const size_t j );
+
+				template< typename T >
```

Comment:
At this point DerivedMatrix is incomplete type and its typedefs are not yet defined. We could add a static check in the function body to check if T is same as DerivedMatrix::access_type. Accessing DerivedMatrix::access_type within function body is fine because function body is "evaluated" after the class is fully "processed".